Adjust how property types are defined to support arrays of all types

tgoyne opened the pull request:
More prep work to support arrays of primitives that should have no functional impact.

The primary change here is to turn `PropertyType::Array` into a bitflag that can be added to any of the other property types, rather than its own type which implies `Object`. Since this makes it so that there's no longer a 1:1 correspondence between `PropertyType` and the core data types, it also goes ahead and renumbers the enum values and also makes the nullability another flag rather than a separate boolean on the `Property`.

This also changes how `is_indexed` and `is_primary` are set on `Property` to use a tagged bool type that makes it a compile-time error to pass them in the wrong order. This change isn't directly related, but untangling it from all the changes to the schema definitions in the tests would have taken a while so it's lumped in.

bdash commented on src/object_schema.cpp:
```diff
@@ -66,6 +53,25 @@ ObjectSchema::ObjectSchema(std::string name, std::initializer_list<Property> per
+        case type_Link:      return PropertyType::Object | PropertyType::Nullable;
+        case type_LinkList:  return PropertyType::Object | PropertyType::Array;
+        case type_Table:     return from_core_type(*table.get_subdescriptor(col), 0) | PropertyType::Array;
+        default: REALM_UNREACHABLE(); // FIXME
```
What remains to be fixed here?

bdash commented on src/object_store.cpp:
```diff
@@ -96,19 +96,35 @@ void add_index(Table& table, size_t col)
+        case PropertyType::Date:   return type_Timestamp;
+        case PropertyType::Data:   return type_Binary;
+        case PropertyType::Object: return type_Link;
+        default: REALM_TERMINATE("unknown type");
```
Is the omission of `Any` intentional?

Is always mapping `Object` to `type_Link` intentional given that it should be `type_LinkList` when the array flag is set?

tgoyne replied:
Mapping to the core type is only needed for creating columns, and it doesn't allow creating Mixed columns. Maybe should explicitly assert that type isn't `Any`.

bdash replied:
On that same vein it seems like omitting `Object` may be sensible since the column creation logic for link columns is different than for other column types.

bdash commented on src/property.hpp:
```diff
@@ -19,145 +19,218 @@
+    Float  = 5,
+    Double = 6,
+    Object = 7,
+    LinkingObjects = 8, // Implies Array
```
What do you mean by "Implies Array" here? Is a linking objects property always be represented as `LinkingObjects | Array`, or is `LinkingObjects` by itself always treated as an array? The implementation of `is_array` suggests the former.

tgoyne commented:
I've now addressed all of the review feedback, rebased this onto master, and added some more tests for schema validation logic (this adds some more illegal states to check, and there were a bunch of existing ones that weren't tested).

bdash commented:
It looks like there's a build failure here due to #434 introducing new references to `Property::is_nullable`.

bdash commented on src/object_accessor.hpp:
```diff
@@ -85,7 +85,24 @@ void Object::set_property_value_impl(ContextType& ctx, const Property &property,
+        REALM_ASSERT(property.type == PropertyType::Object);
+
+        List list(m_realm, m_row.get_linklist(col));
+        list.remove_all();
```
This appears to suffer from the same issue that you fixed in realm/realm-cocoa#5127. It looks like a preexisting issue, so it's not necessary to fix it prior to merging unless it's trivial to do so.

tgoyne replied:
I'll make another PR to fix that in a bit.
